galaxyproject/galaxy
# Update tool recommendation API using transformers (neural network)

This PR updates the tool recommendation API in two ways:
- By adding a new model trained on the latest data (last 5 years' workflows and 1-year tools usage on Galaxy EU). The new model is here: https://github.com/galaxyproject/galaxy-test-data/pull/22
- Update the neural network architecture to a modern transformer-based architecture

The new model achieves better generalization in terms of predictions beyond training data, is faster (needs only half the amount of time compared to the existing RNN-based model) in converging and requires only 1/4 of the time for a prediction.

Some plots:

![model_load_time](https://user-images.githubusercontent.com/3022518/195872550-e5675c04-646d-40b7-a5bf-945935d9ea0f.png)

![training](https://user-images.githubusercontent.com/3022518/195872554-78a8e9e3-3305-4f62-b444-7a74829690f6.png)

Generalisation in prediction:

Let's take a tool: "anndata_import". In the training data (workflows from Galaxy EU), the following tools are following tools of "anndata_import":
scanpy_filter, anndata_inspect, anndata_manipulate, ucsc_cell_browser, scanpy_inspect, scanpy_filter_cells

Prediction by RNN: 'anndata_manipulate', 'scanpy_filter', 'scanpy_filter_cells', 'ucsc_cell_browser', 'scanpy_inspect', 'anndata_inspect', 'scanpy_plot', '**TransTermHP**', '**edu.tamu.cpt.gff3.require_sd**', 'scanpy_normalise_data', '**edu.tamu.cpt.genbank.shinefind**', '**gd_raxml'**, '**mmpbsa_mmgbsa**', '**edu.tamu.cpt.util.wigToBigWig**', '**edu.tamu.cpt2.util.glimmer3_to_gff3**', 'scmap_scmap_cluster', 'scmap_scmap_cell', 'tp_tail_tool', 'scanpy_filter_genes', '**nanocompore_sampcomp**'

Prediction by Transformer: 'scanpy_filter_cells', 'anndata_inspect', 'ucsc_cell_browser', 'scanpy_filter', 'scanpy_inspect', 'anndata_manipulate', 'scanpy_normalise_data', '**MergingDataClusterize**', '**gatk_bqsr**', 'scanpy_plot', 'anndata_ops', 'scanpy_remove_confounders', 'scanpy_integrate_harmony', '**splitTranscriptGff**', 'scanpy_normalize', 'scpred_get_feature_space', 'scanpy_find_variable_genes', 'scpred_predict_labels', '**rdpmulticlassifier**', 'scpred_eigen_decompose'

Both the models correctly predict the "true" tools of "anndata_import" but RNN predicts 8 (out of 20 predicted tools) incorrectly (ie these 8 tools do not belong to the single-cell field) while the Transformer prediction only 4 tools (again out of 20) that do not belong to the single-cell field. Incorrectly predicted tools are shown in bold.

There are many such examples that can be seen for tools or a sequence of tools. Currently, Transformers-based architectures are state-of-the-art for modelling sequences (https://arxiv.org/abs/1706.03762).

## How to test the changes?
(Select all options that apply)
  1. First of all, this PR https://github.com/galaxyproject/galaxy-test-data/pull/22 that enable the new model to be downloaded in Galaxy via its config "tool_recommendation_model_path"
  2. Enable Galaxy config "enable_tool_recommendations" and set to "true"
  3. Enable another Galaxy config "tool_recommendation_model_path"
  4. Also, enable the last Galaxy config "topk_recommendations"
  5. Start the local Galaxy instance and navigate to the workflow editor page
  6. Select any tool and the tool recommendation arrow (top-right of tool's box) is displayed in a dropdown.

![featureC_deseq2](https://user-images.githubusercontent.com/3022518/195877522-7a234b37-6abe-41fd-9d72-0e36329fe873.png)

## License


ping @bgruening 

thank you!
